src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -625,13 +712,22 @@ pub struct CollectionState<T> {
     /// The last reported timestamp bindings, if any.
     /// This is used to differentialize timestamp bindings received before storing them in stash
     pub(super) last_reported_ts_bindings: HashMap<PartitionId, MzOffset>,
+
+    /// The `ShardId` that can be used to read the contents of this collection.
+    pub shard_id: Option<mz_persist_client::ShardId>,
+
+    /// A `ReadHandle` for the backing persist shard/collection. This internally holds back the
+    /// since frontier and we need to downgrade that when the read capabilities change.
+    pub read_handle: Option<ReadHandle<Row, Row, T, Diff>>,
```

Comment:
We should discuss these! It seems like maybe there is a missing abstraction for what there are needed for. Ideally we could bundle this up in an abstraction that presents as what it is needed for, vs "I might need these fields in some cases".

Reply:
Resolved by hiding the ReadHandle, which was causing this behind a trait object. Now there is only one `Option<CollectionReadHandle<>>` field. That will stop being optional once all sources go through persist.

((recording what I responded elsewhere, in Slack))